### ISSUE TITLE
COLLECT_LIST in KSQL returning duplicate results

### ISSUE THREAD

**Issue by AleksandarTokarev** [confluentinc/ksql#10168]
**Describe the bug**
We have a use case where we have one KSQL Table and we have a push query - which queries the table continously and creates/feeds another table from it
```
CREATE or replace table ITEMS_PER_USER_TABLE
    WITH (PARTITIONS = 12, VALUE_FORMAT = 'JSON') AS
SELECT userId, COLLECT_LIST(
        STRUCT("name":=name, "price":=price, "id":=id, "updatedTime"=:updatedTime)) 
        AS items
FROM ITEMS_TABLE
GROUP BY userId
EMIT CHANGES;
```
In the table `ITEMS_TABLE`, we have the items using the `id` as the primary key, while in the new push query, we want to create a table where we will group the items by `userId` - and the result will be a list.
What we have noticed is - when everything works good after setup - there are no issues. But when KSQLDB goes down - we are getting duplicate (stale) items in the `ITEMS_PER_USER_TABLE` (we are not exactly sure why and how this happens).

**To Reproduce**
Steps to reproduce the behavior, include:
1. The version of KSQL.
`0.29`
3. Sample source data. 
In the table `ITEMS_PER_USER_TABLE`  - we have the following items
id,name,price,userId,updatedTime
1,name1,10,1,1703927564000
2,name2,20,2,1703927564001
3,name3,11,1,1703927564002
4,name4,22,2,1703927564003

In the `ITEMS_PER_USER_TABLE` - when all is good - we have the following
userId, items
```
1, [{id=1,name=name1,price=10,updatedTime=1703927564000},{id=3,name=name3,price=11,updatedTime=1703927564002}]
2, [{id=2,name=name2,price=20,updatedTime=1703927564001},{id=3,name=name3,price=22,updatedTime=1703927564003}]
```
But when things go wrong (KSQL being down) - a bug occurs - and duplicate items start piling up (and they dont get cleared when `ITEMS_TABLE` gets more records.
```
1, [{id=1,name=name1,price=10,updatedTime=1703927564000},{id=3,name=name3,price=11,updatedTime=1703927564002},{id=1,name=name1,price=10,updatedTime=1703927565000}]
2, [{id=2,name=name2,price=20,updatedTime=1703927564001},{id=3,name=name3,price=22,updatedTime=1703927564003},
{id=2,name=name2,price=20,updatedTime=1703927566001},{id=3,name=name3,price=22,updatedTime=1703927574003]
```

5. Any SQL statements you ran
```
CREATE or replace table ITEMS_PER_USER_TABLE
    WITH (PARTITIONS = 12, VALUE_FORMAT = 'JSON') AS
SELECT userId, COLLECT_LIST(
        STRUCT("name":=name, "price":=price, "id":=id)) 
        AS items
FROM ITEMS_TABLE
GROUP BY userId
EMIT CHANGES;
```

**Expected behavior**
The function `COLLECT_LIST` should not return duplicate items - because in the previous table the `id` is the key.

**Actual behaviour**
Duplicate items are returned as part of the `COLLECT_LIST`. This should not be the case.  
It is interesting to note that not all the items are being duplicated - sometimes some of them are - sometimes all of them.

**Comment by cadonna** [comment]
@AleksandarTokarev This looks like duplicates that happen due to the at-least-once semantics (ALOS) KSQL guarantees by default. Under ALOS, when a failure happens records might be written multiple times to an aggregate or to an output topic [1]. To solve your issue, you could try to run the application under exactly-once semantics (EOS). I am wondering whether in your specific it would suffice to use `COLLECT_SET` [2] instead of `COLLECT_LIST`. 

[1] https://docs.ksqldb.io/en/latest/operate-and-deploy/exactly-once-semantics/#at-least-once-semantics
[2] https://docs.ksqldb.io/en/latest/developer-guide/ksqldb-reference/aggregate-functions/#collect_set

**Comment by AleksandarTokarev** [comment]
@cadonna Is it possible that the items are being retained with `COLLECT_LIST` ? Because those stale/old items are not in the `ITEMS_TABLE` table?

**Comment by cadonna** [comment]
Now, I realized that `updatedTime` is different for the "duplicates". I am sorry for that!
Could you please post a minimal example that reproduces what you experience?
More specifically:
- the `CREATE`-statement for `ITEMS_TABLE` (BTW, in step 2 above it should be `ITEMS_TABLE`, right?)
- when to restart ksql to reproduce the situation

**Comment by AleksandarTokarev** [comment]
Yes, time `updatedTime` is different for the records, I guess you missed that.
The actual `CREATE` statement above is for the `ITEMS_PER_USER_TABLE` (pasting it below)- I will need to find the one for `ITEMS_TABLE`
```
CREATE or replace table ITEMS_PER_USER_TABLE
    WITH (PARTITIONS = 12, VALUE_FORMAT = 'JSON') AS
SELECT userId, COLLECT_LIST(
        STRUCT("name":=name, "price":=price, "id":=id, "updatedTime"=:updatedTime)) 
        AS items
FROM ITEMS_TABLE
GROUP BY userId
EMIT CHANGES;
```

These `duplicates/retaining` of items occurs when we have KSQLDB issues - and what is more strange to me is that even if the issues are there for few moments - the `stale` records would be gone from `ITEMS_TABLE` - and the actual new records are constantly being updated (therefore the `updatedTime`) - my assumption is that these  `duplicates/retaining` would go away - but they are not - they are staying - but the ones updated are being updated.

**Comment by cadonna** [comment]
At the moment, I do not understand why the `updatedTime` of the duplicates have values that do not exist in the `ITEMS_TABLE`.

**Comment by AleksandarTokarev** [comment]
They do not exist because I have not put them above (and they do not exist in the beginning) - with time the items are being updated in `ITEMS_TABLE`. 
After some time - they exist - but they old ones are gond from the `ITEMS_TABLE`. And the actual main issue that we have is with `ITEMS_PER_USER_TABLE` - because the new ones are there and updating - but we are seeing some stale ones (which in my opinion should not even appear because the `COLLECT_LIST` is collecting the items from the `ITEMS_TABLE`

**Comment by cadonna** [comment]
I see. We need a minimal example to reproduce the situation, otherwise it is just guessing. I tried to repro it but was not able.

**Comment by AleksandarTokarev** [comment]
What else can I provide in order to be more helpful?

**Comment by AleksandarTokarev** [comment]
Regarding the logs in our datadog system that I see for ksql - they are not really helpful - let me know if anything makes sense to you. I can try to look for more logs/data
![image](https://github.com/confluentinc/ksql/assets/6853997/2d9ef92b-f3e3-4d46-a124-5adf6d05f70f)

**Comment by AleksandarTokarev** [comment]
ok @cadonna some more followup on the issues. 
Here are the `CREATE` table statements for the other tables
`ITEMS_TABLE`

```
CREATE OR REPLACE TABLE ITEMS_TABLE WITH (CLEANUP_POLICY=‘compact’, KAFKA_TOPIC=‘ITEMS_TABLE’, PARTITIONS=12, REPLICAS=3, RETENTION_MS=600000) AS SELECT * FROM ITEMS_TABLE_INITIAL EMIT CHANGES;
````

It seems there is one more table that is backing up this `ITEMS_TABLE` - initial one - here is the DDL
`ITEMS_TABLE_INITIAL`

```
CREATE TABLE ITEMS_TABLE_INITIAL (ID STRING PRIMARY KEY, USERID BIGINT, NAME STRING, PRICE DOUBLE, UPDATEDTIME) WITH (KAFKA_TOPIC=‘items.active’, KEY_FORMAT=‘KAFKA’, VALUE_FORMAT=‘JSON’);
```

Also it seems that these issues have occurred when there was OOM in the KSQLDB - here is one of the logs when OOM occurred - 
```
Process _confluent-ksql (pid: 1886306) triggered an OOM kill on itself. The process had reached 2560000 pages in size.
This OOM kill was invoked by a cgroup, containerID: xxxxx.
```

Few things that come on top of my mind:
1. The actual is OOM is concerning - this should not happen in the first place. We might want to increased the RAM
2. Do you think the issue can come from the fact that we are not using the `INITIAL` table - instead we are using the one created from it?

Thanks a lot

**Comment by cadonna** [comment]
@AleksandarTokarev The best thing would be, if you could provide a reproduction with a docker-compose setup. You can find a docker-compose setup that you can re-use under the following link: https://ksqldb.io/quickstart.html#quickstart-content
Once we can reproduce the issue, it gets easier to find the root cause. Otherwise, it is a lot of crystal balling.

**Comment by AleksandarTokarev** [comment]
@cadonna we have added exactly once as a server parameter and it was good for like 2 weeks. Today we upgraded to ARM instances and during the upgrade there was an issue with the volumes/resources and the bug occurred again.
We are thinking of writing our custom UDAF where we would be `dedupping` the LIST in the aggregate part. 
Is there a proper example on how to do that? Or maybe some further hints on how to do it.
